File: packages/fastapi/fastbackend_fastapi/relationship_engine.py
"""Dynamic relationship endpoint creation."""

from __future__ import annotations

from typing import Any, Callable

from fastapi import HTTPException

from fastbackend_fastapi.utils.route_registry import RouteRegistry
from fastbackend_fastapi.utils.pluralize import pluralize


class RelationshipEngine:
    """Creates relationship endpoints dynamically at runtime."""

    def __init__(
        self,
        registry: RouteRegistry,
        store: dict[str, list[dict[str, Any]]],
        relationships: list[dict[str, Any]],
    ) -> None:
        self.registry = registry
        self.store = store
        self.relationships = relationships

    def register_all(self, app: Any) -> int:
        count = 0
        for rel in self.relationships:
            count += self.register_relationship(app, rel)
        return count

    def register_relationship(self, app: Any, relationship: dict[str, Any]) -> int:
        source = relationship["sourceEntity"]
        resource = pluralize(source)
        rel_name = relationship["name"]
        path = f"/{resource}/{{id}}/{rel_name}"
        method = "GET"

        if self.registry.is_overridden(method, path):
            return 0

        handler = self._create_handler(relationship)
        app.add_api_route(path, handler, methods=[method], tags=[source])
        self.registry.register(
            path, [method], entity=source, relationship=rel_name
        )
        return 1

    def _create_handler(self, relationship: dict[str, Any]) -> Callable:
        source = relationship["sourceEntity"]
        target = relationship["targetEntity"]
        rel_type = relationship["type"]
        source_field = relationship.get("sourceField")
        target_field = relationship.get("targetField", f"{source.lower()}_id")

        async def relationship_handler(id: int, limit: int = 20, offset: int = 0) -> list[dict[str, Any]]:
            source_pk = "id"
            source_items = self.store.get(source, [])
            target_items = self.store.get(target, [])

            source_record = next(
                (item for item in source_items if item.get(source_pk) == id), None
            )
            if source_record is None:
                raise HTTPException(status_code=404, detail=f"{source} not found")

            match rel_type:
                case "one-to-many":
                    fk = target_field or f"{source.lower()}_id"
                    related = [item for item in target_items if item.get(fk) == id]
                case "many-to-one":
                    fk = source_field or f"{target.lower()}_id"
                    fk_value = source_record.get(fk)
                    related = [item for item in target_items if item.get("id") == fk_value]
                case "many-to-many":
                    related = target_items[:limit]
                case "one-to-one":
                    fk = target_field or f"{source.lower()}_id"
                    related = [item for item in target_items if item.get(fk) == id]
                case _:
                    related = []

            return related[offset : offset + limit]

        return relationship_handler

File: packages/express/src/engines/relationship-engine.ts
import type { Express, Request, Response, NextFunction } from 'express';
import type { IRRelationship } from '../utils/ir-loader.js';
import type { PrismaStore } from '../utils/prisma-store.js';
import type { RouteRegistry } from '../utils/route-registry.js';
import { pluralize, openApiPathToExpress, toPrismaDelegateKey } from '../utils/path-utils.js';

export class RelationshipEngine {
  constructor(
    private readonly registry: RouteRegistry,
    private readonly store: PrismaStore,
    private readonly relationships: IRRelationship[],
  ) {}

  registerAll(app: Express): number {
    return this.relationships.reduce((count, relationship) => count + this.registerRelationship(app, relationship), 0);
  }

  registerRelationship(app: Express, relationship: IRRelationship): number {
    const source = relationship.sourceEntity;
    const resource = pluralize(source);
    const openApiPath = `/${resource}/{id}/${relationship.name}`;
    const method = 'GET';

    if (this.registry.isOverridden(method, openApiPath)) {
      return 0;
    }

    app.get(openApiPathToExpress(openApiPath), this.createHandler(relationship));
    this.registry.register(openApiPath, [method], { entity: source, relationship: relationship.name });
    return 1;
  }

  private createHandler(relationship: IRRelationship) {
    const sourceDelegate = this.store.getDelegate({ name: relationship.sourceEntity, fields: [] });
    const targetDelegate = this.store.getDelegate({ name: relationship.targetEntity, fields: [] });
    const sourceKey = toPrismaDelegateKey(relationship.sourceEntity);
    const targetKey = toPrismaDelegateKey(relationship.targetEntity);

    return async (req: Request, res: Response, next: NextFunction): Promise<void> => {
      try {
        const id = Number(req.params.id);
        const limit = Number(req.query.limit ?? 20);
        const offset = Number(req.query.offset ?? 0);

        const sourceRecord = await sourceDelegate.findUnique({ where: { id } });
        if (!sourceRecord) {
          res.status(404).json({ detail: `${relationship.sourceEntity} not found` });
          return;
        }

        let related: Record<string, unknown>[] = [];

        switch (relationship.type) {
          case 'one-to-many': {
            const fk = relationship.targetField ?? `${sourceKey}Id`;
            related = await targetDelegate.findMany({
              where: { [fk]: id },
              take: limit,
              skip: offset,
            });
            break;
          }
          case 'many-to-one': {
            const fk = relationship.sourceField ?? `${targetKey}Id`;
            const fkValue = sourceRecord[fk];
            if (fkValue !== undefined && fkValue !== null) {
              const record = await targetDelegate.findUnique({ where: { id: fkValue } });
              related = record ? [record] : [];
            }
            break;
          }
          case 'one-to-one': {
            const fk = relationship.targetField ?? `${sourceKey}Id`;
            const record = await targetDelegate.findUnique({ where: { [fk]: id } });
            related = record ? [record] : [];
            break;
          }
          case 'many-to-many':
            related = await targetDelegate.findMany({ take: limit, skip: offset });
            break;
          default:
            related = [];
        }

        res.json(related);
      } catch (error) {
        next(error);
      }
    };
  }
}
